src/languageSupport/languages/flux/lsp/connection.ts

```diff
@@ -223,27 +261,99 @@ class LspConnectionManager {
+      })
+    }
+    if (toAdd.fields?.length) {
+      toAdd.fields.forEach(value =>
+        this.inject(ExecuteCommand.CompositionAddField, {value})
+      )
+    } else if (toRemove.fields?.length) {
+      toRemove.fields.forEach(value =>
+        this.inject(ExecuteCommand.CompositionRemoveField, {value})
+      )
+    }
+    if (toAdd.tagValues?.length) {
+      toAdd.tagValues.forEach(({key, value}) =>
+        this.inject(ExecuteCommand.CompositionAddTagValue, {tag: key, value})
+      )
+    } else if (toRemove.tagValues?.length) {
+      toRemove.tagValues.forEach(({key, value}) =>
+        this.inject(ExecuteCommand.CompositionRemoveTagValue, {tag: key, value})
+      )
     }
-    // TODO: finish LSP update first
-    // this.inject(ExecuteCommand.CompositionInit, payload)
   }
 
-  _updateLsp(_: SchemaSelection) {
-    // TODO: finish LSP update first
-    // this.inject(ExecuteCommand.Composition<Something>, payload)
+  _diffSchemaChange(schema: SchemaSelection, previousState: SchemaSelection) {
+    const toAdd: Partial<SchemaSelection> = {}
+    const toRemove: Partial<SchemaSelection> = {}
+
+    if (schema.bucket && previousState.bucket != schema.bucket) {
+      toAdd.bucket = schema.bucket
+    }
+    if (schema.measurement && previousState.measurement != schema.measurement) {
+      toAdd.measurement = schema.measurement
+    }
+
+    if (previousState.fields?.length != schema.fields?.length) {
+      toAdd.fields = [] // preserve order
+      const existingFields = new Set(previousState.fields)
+      schema.fields.forEach(f =>
+        !existingFields.has(f) ? toAdd.fields.push(f) : existingFields.delete(f)
+      )
+      toRemove.fields = Array.from(existingFields)
+    }
+
+    if (previousState.tagValues?.length != schema.tagValues?.length) {
+      toAdd.tagValues = [] // preserve order
+      const existingTagValues = new Set(previousState.tagValues)
+      schema.tagValues.forEach(f =>
+        !existingTagValues.has(f)
+          ? toAdd.tagValues.push(f)
+          : existingTagValues.delete(f)
+      )
+      toRemove.tagValues = Array.from(existingTagValues)
+    }
+
+    return {toAdd, toRemove}
   }
 
-  _initComposition(schema: SchemaSelection) {
-    if (!schema.composition.synced) {
-      this._setSessionSync(true)
+  _restoreComposition(schema: SchemaSelection) {
+    this._updateLsp(schema, {})
+  }
+
+  private _i = 0
+  private _o = 0
+  _insertBuffer = req => {
+    this._initBufferComposition[this._i % 100] = req
+    this._i++
+  }
+  _consumeInitBuffer = () => {
+    this._insertIntoBuffer = false
+    let msg = this._initBufferComposition[this._o % 100]
+    while (!!msg) {
+      this.inject(...msg)
+      this._o++
+      msg = this._initBufferComposition[this._o % 100]
     }
+    return
+  }
 
-    this._initLsp(schema)
+  _initComposition() {
+    // handle race condition, on page reloads
+    this._insertIntoBuffer = true
+    setTimeout(() => this._consumeInitBuffer(), 2000)
```

Comment:
This is the race condition, where we are sending executeCommands before the `didOpen` is received.
<img width="342" alt="Screen Shot 2022-08-24 at 11 42 15 AM" src="https://user-images.githubusercontent.com/10232835/186541436-07cece72-1e54-4aa6-a8b2-a54eb1b6fac5.png">

Reply:
note: I did a survey of all the event handlers in the monaco-editor. (On the editor object, and on the model.) There is nothing which fires after the didOpen was initiated.

Hence, the horrible timeout.

Reply:
once we have the middleware (Q4), then we can do this properly by listening to traffic.